src/core_plugins/kibana/public/management/sections/indices/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.js

```diff
@@ -0,0 +1,178 @@
+import {
+  getIndices,
+  isIndexPatternQueryValid,
+  getMatchedIndices,
+  canAppendWildcard,
+  createReasonableWait
+} from '../../lib';
+import { LoadingIndices } from './components/loading_indices';
+import { StatusMessage } from './components/status_message';
+import { IndicesList } from './components/indices_list';
+import { Header } from './components/header';
+
+import {
+  EuiPanel,
+  EuiSpacer,
+} from '@elastic/eui';
+
+export class StepIndexPattern extends Component {
+  static propTypes = {
+    allIndices: PropTypes.array.isRequired,
+    isIncludingSystemIndices: PropTypes.bool.isRequired,
+    esService: PropTypes.object.isRequired,
+    goToNextStep: PropTypes.func.isRequired,
+    initialQuery: PropTypes.string,
+  }
+
+  static defaultProps = {
+    initialQuery: '',
+  }
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      partialMatchedIndices: [],
+      isLoadingIndices: false,
+      query: props.initialQuery,
+      appendedWildcard: false,
+      showingIndexPatternQueryErrors: false,
+    };
+  }
+
+  fetchIndices = async (query) => {
+    const { esService } = this.props;
+
+    this.setState({ isLoadingIndices: true });
+    const partialMatchedIndices = await getIndices(esService, `${query}*`);
+    createReasonableWait(() => this.setState({ partialMatchedIndices, isLoadingIndices: false }));
+  }
+
+  onQueryChanged = (e) => {
+    const { appendedWildcard } = this.state;
+    const { target } = e;
+
+    let query = target.value;
+    if (query.length === 1 && canAppendWildcard(e.nativeEvent.data)) {
+      query += '*';
+      this.setState({ appendedWildcard: true });
+      setTimeout(() => target.setSelectionRange(1, 1));
+    }
+    else {
+      if (query === '*' && appendedWildcard) {
+        query = '';
+        this.setState({ appendedWildcard: false });
+      }
+    }
+
+    this.setState({ query, showingIndexPatternQueryErrors: !!query.length });
+    this.fetchIndices(query);
+  }
+
+  renderLoadingState() {
+    const { isLoadingIndices } = this.state;
+
+    if (!isLoadingIndices) {
+      return null;
+    }
+
+    return (
+      <LoadingIndices/>
+    );
+  }
+
+  renderStatusMessage() {
+    const { allIndices, isIncludingSystemIndices } = this.props;
+    const { query, isLoadingIndices, partialMatchedIndices } = this.state;
+
+    if (isLoadingIndices) {
+      return null;
+    }
+
+    const matchedIndices = getMatchedIndices(allIndices, partialMatchedIndices, query, isIncludingSystemIndices);
+
+    return (
+      <StatusMessage
+        matchedIndices={matchedIndices}
+        query={query}
+      />
+    );
+  }
+
+  renderList() {
+    const { isIncludingSystemIndices, allIndices } = this.props;
+    const {
+      query,
+      partialMatchedIndices,
+      isLoadingIndices,
+    } = this.state;
+
+    if (isLoadingIndices) {
+      return null;
+    }
+
+    const indices = getMatchedIndices(allIndices, partialMatchedIndices, query, isIncludingSystemIndices);
```

Comment:
We're calling getMatchedIndices here and on line 94 and 139. Would it be more efficient to call it once within render() and then pass the result as an argument to renderStatusMessage(), renderList(), and renderHeader()?

Reply:
For sure!